pandas-stubs/core/groupby/grouper.pyi

```diff
@@ -7,67 +14,115 @@ from pandas import (
     Series,
 )
 from pandas.core.groupby.ops import BaseGrouper
+from pandas.core.resample import TimeGrouper
+from typing_extensions import (
+    Self,
+    deprecated,
+)
 
-from pandas._typing import NDFrameT
+from pandas._libs.lib import NoDefault
+from pandas._typing import (
+    ArrayLike,
+    Axis,
+    Frequency,
+    Incomplete,
+    KeysArgType,
+    Level,
+    ListLikeHashable,
+    NDFrameT,
+    npt,
+)
+from pandas.util._decorators import cache_readonly
 
 class Grouper:
-    def __new__(cls, *args, **kwargs): ...
-    key = ...
-    level = ...
-    freq = ...
-    axis = ...
-    sort = ...
-    grouper = ...
-    obj = ...
-    indexer = ...
-    binner = ...
+    key: KeysArgType | None
+    level: Level | ListLikeHashable[Level] | None
+    freq: Frequency | None
+    axis: Axis
+    sort: bool
+    dropna: bool
+    binner: Incomplete
+    @overload
+    def __new__(
+        cls,
+        key: KeysArgType | None = None,
+        level: Level | ListLikeHashable[Level] | None = None,
+        axis: Axis | NoDefault = ...,
+        sort: bool = False,
+        dropna: bool = True,
+    ) -> Self: ...
+    @overload
+    def __new__(cls, *args, freq: Frequency, **kwargs) -> TimeGrouper: ...
     def __init__(
-        self, key=..., level=..., freq=..., axis: int = ..., sort: bool = ...
+        self,
+        key: KeysArgType | None = None,
+        level: Level | ListLikeHashable[Level] | None = None,
+        freq: Frequency | None = None,
+        axis: Axis | NoDefault = ...,
+        sort: bool = False,
+        dropna: bool = True,
     ) -> None: ...
     @property
+    @deprecated("Grouper.ax is deprecated. Use Resampler.ax instead.")
```

Comment:
To make maintaining pandas-stubs easier, we opted to remove methods/arguments that trigger a warning on the current pandas. Unless more people work on pandas-stubs, I would be inclined to keep it simpler by removing these deprecated methods.

(I assume stubtests doesn't like this simplification.)

Reply:
> (I assume stubtests doesn't like this simplification.)

Yes but stubtest can be silenced about a missing symbol once we established the symbol is not to be included in the stubs. There will always be an allowlist of things we intentionally want to deviate from runtime, stubtest can still help with unintentional deviations though.